### ISSUE TITLE
New lint: public API struct became enum or union

### ISSUE THREAD

**Issue by obi1kenobi** [obi1kenobi/cargo-semver-checks#954]
Inspired by this post: https://internals.rust-lang.org/t/rest-patterns-foo-should-match-non-struct-types/21607

TL;DR: Given a type `Example`, the pattern in `matches!(value, Example { .. })` is only valid if `Example` is a struct. It fails on both enums and unions at the moment: [playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=b7fc5c8589620a247846e82e8788c304)

We already have a similar lint for changing a public API struct _with public fields_ into an enum or union: [`struct_with_pub_fields_changed_type`](https://github.com/obi1kenobi/cargo-semver-checks/blob/main/src/lints/struct_with_pub_fields_changed_type.ron). The breakage there is more compelling than "empty pattern match" so we should keep that lint. We should make the new lints disjoint from it, so that users see one or the other but never both.

We also have a lint for constructible structs changing to another type: [`constructible_struct_changed_type`](https://github.com/obi1kenobi/cargo-semver-checks/blob/main/src/lints/constructible_struct_changed_type.ron). The breakage there is also more compelling than "empty pattern match" so we should keep that lint too. The new lints should be disjoint from it as well.

The new lints here should therefore be:
- [x] public API struct with _no public fields_ is no longer a struct: #962 
- [ ] public API `#[non_exhaustive]` struct with no fields of any kind is no longer a struct: #963

The new lints will be quite similar to the related lints mentioned above. Make sure to check out our [contributing guide](https://github.com/obi1kenobi/cargo-semver-checks/blob/main/CONTRIBUTING.md), and take a look at what [prior lints' merged PRs looked like](https://github.com/obi1kenobi/cargo-semver-checks/pull/844) so you know what to expect. It's easier than you think!

Thanks to @jhpratt for bringing this to my attention!
